### DISCUSSION TITLE
qemu学习方法？

### DISCUSSION THREAD

**Discussion by codetang-2417** [QQxiaoming/quard_star_tutorial#5] [Q&A] [answered]
大佬打扰了！我最近正在学习qemu的开发，想问问您有没有好的关于qemu的资料，官方的文档似乎不太合适？比如在用qemu新设计一种cpu的时候，有很多函数的添加，各类宏的使用，还有一些函数的位置，这些我都不太清楚，想知道您是怎么从0搭建起了quard_star？看您的开发日记，大部分都是记录架构，没有提到如何使用和理解qemu

**Comment by QQxiaoming** [comment, marked as answer]
感谢提问。关于qemu的文档资料似乎正如官方的说明，没有足够多的开发者文档。

以下是一些个人见解：

与众多基础开源一样，它们的源码公开，看起来直接阅读源码，跟踪调试会是比较好的学习手段，但是正如qemu这样比较大型的项目直接阅读源码还是有一定困难，我与你相同也是最近一年才开始学习源码的。

最初主要是参考模仿学习hw/arm以及hw/riscv目录下的各类soc实现代码，随着慢慢越发熟悉整个hw里的各种ip和soc实现，我也把[quard_star.c](https://github.com/QQxiaoming/quard_star_tutorial/blob/main/qemu-7.0.0/hw/riscv/quard_star.c)的代码写的更规整易懂了，你也可以参考。

第二就是编译qemu时使用--enable-debug选项，这样可以使用gdb运行qemu，对自己感兴趣的实现代码打断点跟踪，dump数据结构分析源码。

第三就是执行qemu-system-riscv64 -d help可以看到很多trace功能，有助于分析qemu代码本身。

由于我也是多通过查询google，阅读源码进行学习，没有找到很全面的资料。因此只能提供这点经验，以上内容可能无法给你提供实质性的帮助，非常抱歉。

ps：另外你原文“设计一种cpu”我理解的是使用现有的架构如arm/riscv设计soc，而不是设计一款新cpu core架构，如果是要设计新的cpu架构可能需要学习了解qemu中的[tcg](https://wiki.qemu.org/Documentation/TCG)部分。

**Reply by codetang-2417** [reply]
感谢大佬解答，您的解答也很有用，quard_star给了我很多的参考帮助，很佩服你可以从0开始完成这样一个优秀的项目，谢谢！
